### PR TITLE
design: One-pager for Beta Composition Environment, a.k.a. EnvironmentConfigs

### PR DESCRIPTION
**Pull request by phisco**
### Description of your changes

In this one-pager, I attempted to summarize the necessary context to understand the current implementation of the `Composition Environment` (also known as EnvironmentConfigs) as described by the original [one-pager](https://github.com/crossplane/crossplane/blob/master/design/one-pager-composition-environment.md), and to discuss issues that we saw emerging with it.

Please, focus on the above first, so that we can have a shared understanding of what we are trying to solve.

In the end, I try to propose an approach to address the above issues, taking into account the overall direction Crossplane seems to be heading for, and to discuss the next steps we need before promoting this functionality to beta.

Please consider this section as a proposal, I decided to leave it there to have a starting point for discussion, but should be reviewed against any change to the context above and rewritten to be more "prescriptive" with whatever decision we take in the end.

I have: 

- [x] Read and followed Crossplane's [contribution process].
- [ ] ~Run `make reviewable` to ensure this PR is ready for review.~
- [ ] ~Added or updated unit tests.~
- [ ] ~Added or updated e2e tests.~
- [ ] ~Linked a PR or a [docs tracking issue] to [document this change].~
- [ ] ~Added `backport release-x.y` labels to auto-backport this PR.~

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/master/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/master/contributing#checklist-cheat-sheet

### CODE REVIEW

File: design/one-pager-composition-environment-beta.md
Comment (phisco) [reply design/one-pager-composition-environment-beta.md:9]: I'll move the old one to defunct once this one gets approved 👍

File: design/one-pager-composition-environment-beta.md
Comment (phisco) [reply design/one-pager-composition-environment-beta.md:218]: In this case I mean more like the group of maintainers, I'll make it more explicit.

File: design/one-pager-composition-environment-beta.md
Comment (phisco) [reply design/one-pager-composition-environment-beta.md:276]:
```suggestion
Referencing the `environment` twice in `spec.environment.environmentConfigs` feels
[redundant](https://github.com/crossplane/crossplane/issues/4591), reinforcing the confusion between the `in-memory environment` and `EnvironmentConfigs`.
```

File: design/one-pager-composition-environment-beta.md
Comment (phisco) [reply design/one-pager-composition-environment-beta.md:286]:
```suggestion
To the best of our knowledge, it is mostly power users who are selecting and merging
```

File: design/one-pager-composition-environment-beta.md
Comment (phisco) [reply design/one-pager-composition-environment-beta.md:341]: everything else above is conditional, so I'd say:
```suggestion
    - the concept of `in-memory environment` would become just a convention between functions, relying on the [`apiextensions.crossplane.io/environment`](http://apiextensions.crossplane.io/environment) key in the context, as `function-patch-and-transform` already does.
```

However I wouldn't put much focus on the form for this last part, as I want to rewrite it once the above context settles and we agree this is the way to go.

File: design/one-pager-composition-environment-beta.md
Comment (phisco) [reply design/one-pager-composition-environment-beta.md:346]: This was voluntarily left as a question. It's to be discussed whether we care about preserving such behavior first and then how to do that. 
Again don't focus on the form for this section, it's more like a draft proposal to start the discussion as I said in the description of the PR.

Comment (jeanduplessis) [reply design/one-pager-composition-environment-beta.md:349]: You mean I was actually meant to read the description of the PR ;-)

File: design/one-pager-composition-environment-beta.md
Comment (lsviben) [reply design/one-pager-composition-environment-beta.md:9]: Should we move it to defunct right away or when this one is "implemented"? 

Based on our [conventions](https://github.com/crossplane/crossplane/blob/a53d5f1ff9e0f50e44b421ce5a811251a314437c/design/README.md), I believe the old one should stay `Accepted` with a note that this one supersedes it until its fully implemented, then it should go to `Defunct`? I think this would go into the "semi-defunct" category described at the bottom.

**Review by lsviben — COMMENTED** [review]
Thanks @phisco for the great summary of the feature and future plans :raised_hands: . Added a few questions

File: design/one-pager-composition-environment-beta.md
Comment (phisco) [reply design/one-pager-composition-environment-beta.md:9]: yes, I think it makes sense to keep it as semi-defunct, as the parts about `EnvironmentConfigs` will still apply. I'll see what we did in the past and for similar cases and do the same.

File: design/one-pager-composition-environment-beta.md
Comment (phisco) [reply design/one-pager-composition-environment-beta.md:232]: Yes, exactly. It could actually be any known number of resources, the important bit is actually whether you know the resources beforehand or the actual resource to read is controlled by the end user of the composite resource. 

If you know them, today you can create 1 or more EnvironmentConfigs with the needed information and select the right one according to the user's input. 

If you don't know them, you can only compose an ObserveOnly resource according to the user input and therefore create one for each composite resource. 

I'll rephrase it to "Do you need to reference a finite set of Resources you know beforehand?", makes sense?

File: design/one-pager-composition-environment-beta.md
Comment (phisco) [reply design/one-pager-composition-environment-beta.md:348]: sure! `function-patch-and-transform` today only knows about the environment, `function-set-environment` was just an example to make it really explicit it's tied to that, but I would definitely prefer to have it generic from the beginning too 👍
```suggestion
- Selecting and merging EnvironmentConfigs:
    - selection would be done by a dedicated function, e.g. `function-select-extra-resources`.
    - merging into the environment would be done by another dedicated function, e.g. `function-set-environment` or a more generic `function-set-context`.
```

File: design/one-pager-composition-environment-beta.md
Comment (lsviben) [reply design/one-pager-composition-environment-beta.md:232]: thanks for the clarification. 

I like "Do you need to reference a finite set of Resources you know beforehand?". 

Some other suggestion:
"Do you need to reference a finite set of already known Resources?"

But your suggestion is good as well.

File: design/one-pager-composition-environment-beta.md
Comment (turkenh) [inline design/one-pager-composition-environment-beta.md:356]: Just catching up with the latest state there. I need some time to process but my early thoughts are:

I am not yet convinced that the approach there should be the way we support composition environments while it might make sense for requesting extra resources. I feel like we need a more "native" support for environments rather than supporting them as a side effect (of capability of requesting extra resources) with a good amount of boilerplate code and 1-2 function calls each responsible with some part. This is mostly based on the fact that we know composition environments are crucial for any non-trivial composition and the desire to provide a smooth UX to enable this. So, my mind still goes to a solution like [this](https://github.com/crossplane/crossplane/issues/4583#issuecomment-1732343924) instead of splitting this functionality to dedicated functions.

If I try to make an analogy:

We are designing UNIX and discussing how someone should pass some variables into commands and instead of defining/inventing the thing we know as Environment Variables in UNIX, we are talking about suggesting something like the following as a solution for those who need such a functionality:

```shell
env-requester | env-merger | cmd1 | cmd 2
```

instead of:

```shell
export VAR1=value1
source /path/to/somefile
cmd1 | cmd2
```

I know this is not the perfect example, so please don't list the differences between a UNIX environment and K8s/Crossplane etc, the point I am trying to make is more about embracing a widely needed functionality and providing the best possible UX with native support vs "enabling" it somehow.

**Review by turkenh — COMMENTED** [review]
Great write up @phisco 🙌 

I agree that it is very important that we are on the same page regarding what we are trying to solve before attempting to solve it 🙂 One thing that is not clear to me though, do you want to merge this without having the actual design or do you see this as a living document and plan to get it merged after adding details  and once we reach to a consensus on the final design? Because I am not sure if it makes sense to aim merging it as a design doc without having the actual design 🤔

**Comment by phisco** [comment]
@turkenh, I do want to rewrite the final part before merging, although it might be more a "plan" rather than a "design", as if we agree on the one proposed, the actual design will be needed for https://github.com/crossplane/crossplane/issues/4739, through its own one-pager, which I'm already working on; while here we'll only say we'll rely on that and take note of the needs we should take into consideration while designing that.

File: design/one-pager-composition-environment-beta.md
Comment (phisco) [reply design/one-pager-composition-environment-beta.md:228]: Actually I would rewrite that as:
```
Currently, a Composition author who wanted to address some information external
to the Composite and Composed Resources:
```
the point is that they are "external to the composition", not really whether they are kubernetes or cloud resources. WDYT?

File: design/one-pager-composition-environment-beta.md
Comment (phisco) [reply design/one-pager-composition-environment-beta.md:246]: what do you mean by `compiled`?

File: design/one-pager-composition-environment-beta.md
Comment (phisco) [reply design/one-pager-composition-environment-beta.md:356]: As per your example, the two things don't exclude each other, in shell scripts, piped commands have access to the environment.

To continue with the UNIX analogy, I see enabling functions to retrieve resources more like providing UNIX sockets and leaving to users to decide what to do with them. 

What I'm suggesting is indeed, given that we have a finite amount of time, to put on hold the development of a native approach and instead focus on enabling functions. Nothing would block us from "bringing back in" as native the "best possible UX" once it emerges from functions and we agree on it.

File: design/one-pager-composition-environment-beta.md
Comment (phisco) [reply design/one-pager-composition-environment-beta.md]: I rephrased it as above, let me know if there is still something that can be clarified.

File: design/one-pager-composition-environment-beta.md
Comment (phisco) [reply design/one-pager-composition-environment-beta.md]: I removed the question and moved to the US syntax 😉

**Comment by phisco** [comment]
FYI, the proposal to add extra resource gathering capabilities to Functions: https://github.com/crossplane/crossplane/pull/5099

**Comment by phisco** [comment]
I think we reached consensus on the path forward here, see the last part of the proposal.

I understand that it's a weird Design Document. It's not actually proposing a design; it's more of an issue statement/documentation and a decision log. We clearly state that we won't be actively investing further into the native "Composition Environment" and won't set any timeline for its promotion to Beta. This might never happen if the function-based approach we think could replace the in-tree implementation works (NOTE: in such a case, we would offer as smooth a transition as possible from the native approach).

So, I still think it might be worth merging this just to have something to point people at when they ask when Composition Environment (a.k.a. EnvironmentConfigs) is going to be promoted to Beta or more trivial questions like "how to patch back to the EnvironmentConfig".

WDYT?
CC @MisterMX @negz

**Comment by negz** [comment]
@phisco Yeah we should definitely merge it.

It feels more like an ADR, which is something we wanted to adopt but never did https://github.com/crossplane/crossplane/issues/3602.

Maybe we could rename this to `design/adr-00-???.md` and merge it?

**Comment by phisco** [comment]
@negz, although I do like the idea in theory, in practice, I'm not a huge fan of ADRs. In my experience, they require a lot of discipline to get right, and it feels like there is a lot of overlap with the context already available in PRs and commits. But, maybe I just never saw them used as they should be; that's totally plausible.

I feel this could still be called a "one-pager" as I think it fits the definition provided by the `design/README.md`:
```
   A One-pager is a brief document pitching an idea for further
   experimentation. One-pagers, as the name suggests, are usually around one
   page long. They provide just enough context to achieve buy-in from Crossplane
   maintainers.
```

But the name of the file is definitely not a hill I would die on, so I could go either way.

**Review by negz — APPROVED** [review]
@phisco I don't feel super motivated to take on ADRs, so I'm happy to merge as is. I forget we'd defined what a one-pager was for, and I agree this does seem to fit within those bounds.

**Pull request merged**
